immutablecollections/_immutableset.py

```diff
@@ -28,6 +29,93 @@
+    platform.python_version_tuple() >= ('3','6','0') and
+    platform.python_implementation() == "CPython"
+)
+
+
+def immutableset(iterable: Optional[Iterable[T]]=None,
+                 *,
+                 disable_order_check: bool =False) -> 'ImmutableSet[T]':
+    """
+    Create an immutable set with the given contents.
+
+    The iteration order of the created set will match *iterable*.  Note that for this reason,
+    when initializing an ``ImmutableSet`` from a constant collection expression, prefer a
+    list over a set.  We attempt to catch a handful of common cases in which determinsitic
+    iteration order would be discarded: initializing from a (non-``ImmutableSet``) set or
+    (on CPython < 3.6.0 and other Pythons < 3.7.0) initializing from a dictionary view. In
+    these cases we will throw an exception as a warning to the programmer, but this behavior
+    could be removed in the future and should not be relied upon.  This check may be disabled
+    by setting *disable_order_check* to ``True``.
+
+    If *iterable* is already an ``ImmutableSet``, *iterable* itself will be returned.
+    """
+    # immutableset() should return an empty set
+    if iterable is None:
+        return _EMPTY
+
+    if isinstance(iterable, ImmutableSet):
+        # if an ImmutableSet is input, we can safely just return it,
+        # since the object can safely be shared
+        return iterable
+
+    if not disable_order_check:
+        # case where iterable is an ImmutableSet (which is order-safe)
+        # is already checked above
+        if isinstance(iterable, AbstractSet):
+            raise ValueError("Attempting to initialize an ImmutableSet from "
+                             "a non-ImmutableSet set. This probably loses "
+                             "determinism in iteration order.  If you don't care "
+                             "or are otherwise sure your input has determinstic "
+                             "iteration order, specify disable_order_check=True")
+        # we put this check on the outside to avoid isinstance checks on newer Python versions
+        if not (_python_version_guarantees_deterministic_dict_iteration or
+                _python_implementation_has_deterministic_dict_iteration):
+            if isinstance(iterable, KeysView) or isinstance(iterable, ValuesView) or isinstance(
+                    iterable, ItemsView):
+                    raise ValueError("Attempting to initialize an ImmutableSet from "
+                                     "a dict view. On this Python version, this probably loses "
+                                     "determinism in iteration order.  If you don't care "
+                                     "or are otherwise sure your input has determinstic "
+                                     "iteration order, specify disable_order_check=True")
+
+    # iterable might not be safe to iterate multiple times, so we copy it
+    values = tuple(iterable)
```

Comment:
I think we don't need the tuple conversion, since we only iterate once.

Reply:
Yep, this a relic of the type checking option, where we do need to iterate twice. But when/if the type checking is implemented, we can just do type checking on the constructed object and avoid the tuple-ization.